opentelemetry/proto/collector/metrics/v1/metrics_service.proto

```diff
@@ -42,4 +42,18 @@ message ExportMetricsServiceRequest {
 }
 
 message ExportMetricsServiceResponse {
+  // The details of a partially successfull export request.
+  // When the request is only partially successfull (due to e.g. invalid/bad data),
+  // the server MUST populate the details property with a developer-facing error
+  // message along with the number of accepted data points.
+  ExportMetricsPartialSuccess details = 1;
+}
+
+message ExportMetricsPartialSuccess {
+  // The number of accepted data points
+  int64 accepted_data_points = 1;
+
+  // A developer-facing human-readable error message in English. It should
+  // both explain the error and offer an actionable resolution to it.
+  string error_message = 2;
```

Comment:
This makes it hard for the client to perform any systematic handling besides putting the error message to a log file or somewhere else.

Consider something like:

```json
{"accepted_data_points": 5, "error": [
  {"index": 0, "status_code": 400, "message": "the timestamp xyz is invalid"},
  {"index": 3, "status_code": 502, "message": "the tenant does not exist"}
]}
```

Reply:
Very much agreed. The response should be very clear in indicating:
* which records were accepted successfully
* which records should be re-sent (e.g. after a throttling timeout)
* which should be dropped forever (e.g. due to the user having no entitlement or data being invalid).

Reply:
I do not think that the detailed, structured error response is always desirable. If my receiver is under stress and needs to drop a significant number of records I probably do not want to spend time constructing a large payload with potentially thousands of elements in the error field in the response.

We may want the additional, more detailed "error" field that include this information, but I think it must be optional. I think a simple error_message needs to stay and is the one that is always required.

I would also prefer to start small, with just error_message first and only add the more detailed response later, if we really see the need and clearly understand the use cases.

Reply:
> We also need to consider client side retry - need to let the client know which part of data should be retried (e.g. in a partial success, there could be some retriable error AND some non-retriable error).


I'm not sure if this is even possible on the client. I feel there's too many variables involved and would be quite tricky to get retry in place. For example, back-ends have different requirements on how they deal with metrics and what they allow (e.g. special chars in metric names, number of attributes, etc) how would we make this generic enough that clients could figure it out what is "retry able" and what is not? Then there's things like quota exceeded, authorization (my subscription does not allow me to ingest certain types of logs).. it's just a lot, and to me, very close to back-end things and not OTel.

The [current spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#bad-data) says that for "Bad Data" the client must not retry and I feel for partial success is the same.  I'd vote to keep retry out of the equation here.

Reply:
> I do not think that the detailed, structured error response is always desirable.

+1, it should be optional instead of required.

Reply:
I opened an issue to discuss the retry + more structure #404 to not block this. I will mark this as resolved.